### DISCUSSION TITLE
在有道词典笔三代上实现观看视频【搬运】

### DISCUSSION THREAD

**Discussion by WINFMEMZ2020** [PenUniverse/PenMods#184] [奇技淫巧]
也不能说是【搬运】，但是此方法来自原PenMods社区，原作者Aronnax。在此基础上进行了一些探索，用自己的话再描述一遍。
但原来的社区的数据没了，所以我再发一次，希望能够有帮助。

将视频文件拷如词典笔的Music目录下，比如我在Music目录下新建了一个文件夹叫Video，在Video文件夹内拷入视频222.mp4，视频链接就是 file:///userdisk/Music/Video/222.mp4，其它以此类推。记住这个地址。

将词典笔连上电脑，登录adb，使用以下命令将数据库导出：
```
adb pull /userdisk/math/exerciseFav/exerciseFavorite.db exerciseFavorite.db
```
然后得到exerciseFavorite.db，使用SQLiteStudio打开，主要是要编辑表table_knowledge和表table_mathexercise_ursphoneyxxxxxxx
![该图的表table_knowledge和表table_mathexercise_ursphoneyxxxxxxx就是编辑对象](https://github.com/PenUniverse/PenMods-release/assets/64178216/2b233521-fb36-4553-a96b-c462b89f3465)
先打开表table_mathexercise_ursphoneyxxxxxxx，切换到“数据”选项卡，新建一行数据，按照以下格式填写（一共有24列）
```text
code:随便填
type：2
text：题目，随便填
ques_body_type:1-3随便填
ques_body_content:题目内容，随便填
pattern_type:NULL
pattern_content:NULL
answer_type:0-3随便填
answer_content:随意
analysis_type：0-3随便填
analysis_content：留空或NULL
scan_image_url：NULL
scan_image_local：NULL
knowledge：
{"knowId":1164,"knowName":"Test123","knowVideo":"https://ydhardwarebusiness.nosdn.127.net/b70c028ded3d78c0435a31460247e73d.mp4"}
其中knowId随意（1164即可），knowName为“当前知识点”的名称，knowVideo为知识点视频，这些都是在题目底下的信息。
explain_video：NULL
point_video：NULL
knowledgeId：随便的数字，不能与已存在的冲突
这个很重要！
simQuesList：[]
这个就是相似问题的部分，如果要有内容就这样：
[
    {
        "mathId": "13b67581280235b5cbd91dffba3da4986bc1a",
        "mathType": 3,
        "text": "求比值。（填分数） 26 : 52=___ 16t : 14t=___ 27cm : 63cm=___"
    }
]
label：题目标签，填入[]即可，
如需内容就[ "思路错误"]
source：dictpen
newContent：0
item_state：1
sync_state：0或1
timestamp：时间轴，会影响到题目显示的加入日期与排序，随便填，单位为毫秒，这个也很重要
```
提交数据库（上面的绿色的打勾符号），这一步相当于保存。
然后再到表table_knowledge中新建一行，然后按照如下填写:
```
code：刚刚填写的knowledgeId
text：希望显示在”知识点清单“的名称
video：视频位置
parents：视频参数，[
    {
        "knowId": 3745,
        "knowName": "反比例的意义",
        "knowVideo": "file:///userdisk/Music/Video/222.mp4"
    }
]
children：[]
simQuesList:相似问题列表,直接填[]就可以，但要有内容的话
[
    {
        "mathId": "13b6758133562f98ed5bc3f6c862609372b59",
        "mathType": 3,
        "text": "长方形的面积一定，长和宽（    ）。 A. 成正比例关系 B. 成反比例关系 C. 不成比例"
    },
    {
        "mathId": "13b67581335632a06b8d5349051f30e0cc345",
        "mathType": 3,
        "text": "工作总量一定，已完成的量和未完成的量成反比例关系。"
    },
    {
        "mathId": "13b6758133564e4dc63331981e0ad1d91b6ad",
        "mathType": 3,
        "text": "用一块橡皮泥捏不同的圆柱体，圆柱体的底面积和高（    ）。 A. 成正比例关系 B. 成反比例关系 C. 不成比例"
    }
]
sync_state：填入1
timestamp：填写刚刚在tabel_mathexercise_user:写的timestamp
```
别忘了提交数据库！

在添加自己的视频时，需要修改表table_knowledge中的video和parents中的knowVideo。最后，表table_knowledge中的text为显示在知识点清单中的名称，可以自定义。
视频文件的名称不要包含中文及符号，最好只有英文和数字，不然播放视频会卡在加载中；其它很多原因也会导致卡在加载中，需要自己慢慢试。
除了code和timestamp不能重复外，其它都可以重复，所以复制粘贴即可。
![image](https://github.com/PenUniverse/PenMods-release/assets/64178216/514f642f-4d44-4ed7-a189-9cddec85e391)

最后将数据库复制回词典笔：
```
adb push exerciseFavorite.db /userdisk/math/exerciseFav/exerciseFavorite.db
```

然后还有一些注意事项：
1.视频在播放前会判断联网状态，即便是本地视频，所以暂时不能作为离线播放器来使用。但是这个仅仅是判断联网状态，空壳WiFi（无法访问互联网）也能正常播放。
![邮箱截图](https://github.com/PenUniverse/PenMods-release/assets/64178216/79a2887a-ae3a-431b-aaf7-f470467889cb)
2.视频文件为mp4文件，且文件不能太大，不然播放视频会卡死，只能强制重启解决。右上角倍速也最好不要动，试过点了之后有概率卡死。目前拷入的视频每集大概在280MB，能够正常播放。
3.与单词本相同，在词典笔账号退出（包括cookie过期）后好题本数据会清空，重新登录原来的账号即可恢复。

希望大家都能在词典笔上成功播放视频，也能充分利用以下32G或16G空间（笑），但是这块屏幕色域和色准都比较糟糕（可能是我看的问题），实测三代专业版不能双指放大（不支持多点触控）但播放器有提示。

==========
还有广告（误

最近批量增添视频的脚本最近写好了 [Github](https://github.com/WINFMEMZ2020/YoudaoDictpen_SQLEditor) 在右边的Releases就能看到打包好的程序了。直接新建一个文件夹用来存放要导入词典笔的视频，再新建一个文件夹存放处理好的视频（文件夹路径最好不要有中文和符号）。然后将词典笔连上电脑，打开adb调试（设置-关于-法律监管-3C标识点20下）。打开下载好的release包（解压的路径也最好不要有中文），解压打开里面的luncher.exe(打的不是launcher是故意的)，按照提示填好四个文本框，然后按照软件提示复制视频、传输数据库就能批量导入视频了。
如果卡在了闪光标或提示密码错误说明下错了版本，换一个版本，此程序区分3代和X3S版本（没用过X3S，说是这个版本纯粹是根据adb密码瞎猜的）。

==========
写的好乱的感觉，希望能看懂吧。。。

**Comment by AronnaxM** [comment]
PM社区原帖作者前来报道，感谢转载，说的比我当时清楚多了，在此提出一些问题：
在后续测试的时候，对于联网判定方式仍然存疑，例如就算连接用esp32/82666造出来的空壳WiFi，或者连接某些智能家电的自带WiFi都过不了联网检测，在使用WireShark抓包时发现打开AI好题本时词典笔连接了一堆网站，里面居然还有rom.miui.com（真的假的我都怀疑抓错包了），所以提出一种可能性，也许是否联网是通过与特定网站握手返回一个特定的结果来确认的，也就有了所以三角形是正方形测试的就算连着校网也能打开视频；以及在后续和所以三角形是正方形讨论的过程中发现数据库超过40条就不显示的问题。
另外对于楼主说的不能双指放大的问题，我这里型号是YDP038 CHN PRO，也是三代专业版，为什么我是支持双指放大的捏 ¯\_(ツ)_/¯

**Reply by WINFMEMZ2020** [reply]
嗯，对于网络判断确实也不是很明白，没有测试过，所以就把邮箱里的截图放过来了。但抓到rom.miui.com可能是miui的一个wifi连通性测试网站http://connect.rom.miui.com/generate_204
，说明还可能是有连通性判断的🤔。
另外，上限40条的问题我这里不会出现，我通过脚本导入了明显不止40条了，一百多条都能显示出来。
另外，我的设备不支持双指缩放，型号为YDP031 CHN PRO。YDP038开头的型号应该是X3S（查的），和3代还是有一定区别的（X3S发布比3代晚）。这我也搞不清楚，有道的词典笔发的版本太多了，也有可能是买的比较早的问题。

**Reply by Galaxy-YCY** [reply]
为啥我的三代专业版无法安装PM，显示没有对应版本的PM可供安装，求助

**Reply by WINFMEMZ2020** [reply]
因为pm还没有适配3代

**Reply by Galaxy-YCY** [reply]
你的不是三代吗，是哪个型号啊

**Reply by Eganchiyu** [reply]
回来社区报道力！
当时没想那么多，后来也确实遇到连上wifi没通验证。所以那张截图其实是不正确的。（我当时确实很激动，而且学校网如果不用账号密码登录，确实没有网络下行。）
现在我觉得还是看ping的网站有没有反应吧？
大佬[AronnaxM](https://github.com/AronnaxM)还是很细致的，当时纠正了我的认知
